src/fiber/fiber.ml

```diff
@@ -666,6 +666,40 @@ module Mvar = struct
         k ())
 end
 
+module Svar = struct
+  type 'a t =
+    { mutable current : 'a
+    ; mutable waiters : ('a K.t * ('a -> bool)) list
+    }
+
+  let read t = t.current
+
+  let wait t ~until k =
+    if until t.current then
+      k t.current
+    else
+      let wait = K.create k in
+      t.waiters <- (wait, until) :: t.waiters
+
+  let create current = { current; waiters = [] }
+
+  let write t a k =
+    t.current <- a;
+    let sleep, awake =
+      List.rev_partition_map t.waiters ~f:(fun (k, f) ->
+          if f t.current then
+            Right k
+          else
+            Left (k, f))
+    in
+    (match awake with
+    | [] -> ()
+    | awake ->
+      t.waiters <- List.rev sleep;
+      List.iter awake ~f:(fun k -> K.enqueue k a));
+    k ()
```

Comment:
Would it make sense to call `k` before awaking the readers? It seems like it would make sense for an Svar to prioritize the writer.

Reply:
At first glance, the behaviour of `Svar.write` and `Ivar.fill` should be the same to avoid confusion, i.e., first wake up the readers, then continue.

Why do you think prioritising the writer may be better?

I like to think about `Svar` as just `Ivar` that can be used more than once.

Reply:
I don't remember anymore why we wake up the readers with the Ivar first, but even there it seems like it would make sense to resume the caller of `fill` first (at least to me).

> Why do you think prioritising the writer may be better?

The main motivation for introducing this was to have a wait of notifying many readers without ever blocking the writer. So prioritizing the writer goes along with that idea.

Reply:
I see. Never blocking the writer seems like a nice property to have, indeed. I'm also unsure why `Ivar`s are implemented the way they are at the moment.

Reply:
The current code doesn't block the writer. The `List.iter awake ~f:(fun k -> K.enqueue k a))` only add jobs to the job queue, it doesn't run anything. `k ()` needs to come last to preserve tail-recursion.

Reply:
Ah, indeed, that makes sense!